sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala

```diff
@@ -56,6 +56,28 @@ trait TimeZoneAwareExpression extends Expression {
   @transient lazy val zoneId: ZoneId = DateTimeUtils.getZoneId(timeZoneId.get)
 }
 
+trait TimestampFormatterHelper extends TimeZoneAwareExpression {
+
+  protected def formatString: Expression
+
+  protected def isParsing: Boolean
+
+  @transient final protected lazy val formatterOption: Option[TimestampFormatter] =
+    if (formatString.foldable) {
+      Option(formatString.eval()).map(fmt => getFormatter(fmt.toString))
+    } else None
+
+  final protected def getFormatter(fmt: String): TimestampFormatter = {
+    val formatter = TimestampFormatter(
+      format = fmt,
+      zoneId = zoneId,
+      legacyFormat = SIMPLE_DATE_FORMAT,
+      needVarLengthSecondFraction = isParsing)
+    formatter.validatePatternString()
```

Comment:
do we need this? It's already done in `TimestampFormatter.apply` AFAIK

Reply:
`TimestampFormatter.apply` works for the new Parser.  Adding this to fail the legacy formatter here will give us a narrower scope of influence.

Reply:
can we call `validatePatternString` in `TimestampFormatter.apply` for legacy parser as well?

Reply:
Let me push a commit to verify with jenkins

Reply:
this looks ok through Jenkins